### PR TITLE
maxBytesPerFile read check off-by-one

### PR DESCRIPTION
**Pull request by xkeyideal**

### CODE REVIEW

**Comment by xkeyideal** [comment]
#14

**Comment by xkeyideal** [comment]
@ploxiln 

Please see the newest pr code.

use curReadFileSize instead of maxBytesPerFile for move to next file.

`TODO: each data file should embed the maxBytesPerFile as the first 8 bytes (at creation time) ensuring that the value can change without affecting runtime` be solved.

File: diskqueue.go [hunk 1]
```diff
@@ -625,7 +644,16 @@ func (d *diskQueue) ioLoop() {
 			count = 0
 		}
 
-		if (d.readFileNum < d.writeFileNum) || (d.readPos < d.writePos) {
```
Comment (ploxiln) [inline diskqueue.go:628]: this expression did not need to change (and the original is more concise)

File: diskqueue.go [hunk 2]
```diff
@@ -326,10 +340,10 @@ func (d *diskQueue) readOne() ([]byte, error) {
 	d.nextReadPos = d.readPos + totalBytes
 	d.nextReadFileNum = d.readFileNum
 
-	// TODO: each data file should embed the maxBytesPerFile
-	// as the first 8 bytes (at creation time) ensuring that
+	// use the read file exact size, ensuring that
 	// the value can change without affecting runtime
-	if d.nextReadPos > d.maxBytesPerFile {
+	// only readFileNum less than writeFileNum need move next file
+	if d.readFileNum < d.writeFileNum && d.nextReadPos >= d.curReadFileSize {
```
Comment (ploxiln) [inline diskqueue.go:346]: I'm not sure if `d.readFileNum < d.writeFileNum` needed to be added here, I'll think on it a bit...

Adding a test or two would help :) In particular, one that fails with current master, and passes with this branch. Note that Travis-CI currently fails to post "commit status" for the "legacy travis-ci.org integration". I've updated a couple of other repos to use GitHub Actions instead, I'll get around to this one eventually. In the meantime, you can view test results at https://travis-ci.org/github/nsqio/go-diskqueue/pull_requests or run them locally.

**Comment by ploxiln** [comment]
The new `curReadFileSize` logic looks good to me, thanks!
It will take some more time to review and validate, sorry.

File: diskqueue.go (hunk 2 above)
Comment (xkeyideal) [reply diskqueue.go:346]: when `d.readFileNum == d.writeFileNum` means reader and writer is same file, so when `d.nextReadPos >= d.curReadFileSize` the readFile can't be closed, the program should be waiting writeFile update  `d.curReadFileSize`

And I think just pass now unit test will be ok.

**Comment by xkeyideal** [comment]
I use the `nsqio/go-diskqueue` code develop read-write separation program, for readonly and read by write, thanks for your code.

I use `curReadFileSize` to solve switch files avoid `maxBytesPerFile` cause file data error

File: diskqueue.go (hunk 1 above)
Comment (xkeyideal) [reply diskqueue.go:628]: Yes, the change just for my debug, and forget change back.
I will change it back, and submit another pr.

Comment (ploxiln) [reply diskqueue.go]: no need for another PR: you can do a git interactive-rebase, and force-push this branch, to make any change to this PR

**Comment by mreiferson** [comment]
The logic here looks good and addresses an outstanding TODO, I've cleaned things up in #15, so let's wrap this up there.

Thanks!

**Pull request closed without merging**